autonomi/src/uploader/upload.rs

```diff
@@ -0,0 +1,1235 @@
+            })
+            .collect();
+    }
+
+    loop {
+        // Break if we have uploaded all the items.
+        // The loop also breaks if we fail to get_store_cost / make payment / upload for n consecutive times.
+        if uploader.all_upload_items.is_empty() {
+            debug!("Upload items are empty, exiting main upload loop.");
+
+            // To avoid empty final_balance when all items are skipped. Skip for tests.
+            #[cfg(not(test))]
+            {
+                uploader.upload_final_balance = uploader
+                    .wallet
+                    .balance_of_tokens()
+                    .await
+                    .inspect_err(|err| {
+                        error!("Failed to get wallet balance: {err:?}");
+                    })?;
+            }
+
+            #[cfg(test)]
+            trace!("UPLOADER STATE: finished uploading all items {uploader:?}");
+            let summary = UploadSummary {
+                storage_cost: uploader.tokens_spent,
+                final_balance: uploader.upload_final_balance,
+                uploaded_addresses: uploader.uploaded_addresses,
+                uploaded_count: uploader.uploaded_count,
+                skipped_count: uploader.skipped_count,
+                uploaded_registers: uploader.uploaded_registers,
+            };
+
+            if !uploader.max_repayments_reached.is_empty() {
+                error!(
+                    "The maximum repayments were reached for these addresses: {:?}",
+                    uploader.max_repayments_reached
+                );
+                return Err(UploadError::MaximumRepaymentsReached {
+                    items: uploader.max_repayments_reached.into_iter().collect(),
+                });
+            }
+
+            return Ok(summary);
+        }
+
+        // try to GET register if we have enough buffer.
+        // The results of the get & push register steps are used to fill up `pending_to_get_store` cost
+        // Since the get store cost list is the init state, we don't have to check if it is not full.
+        #[cfg(feature = "registers")]
+        {
+            while !uploader.pending_to_get_register.is_empty()
+                && uploader.on_going_get_register.len() < uploader.cfg.batch_size
+            {
+                if let Some(reg_addr) = uploader.pending_to_get_register.pop() {
+                    trace!("Conditions met for GET registers {:?}", reg_addr.xorname());
+                    let _ = uploader.on_going_get_register.insert(reg_addr.xorname());
+                    interface.submit_get_register_task(
+                        uploader.client.clone(),
+                        reg_addr,
+                        task_result_sender.clone(),
+                    );
+                }
+            }
+        }
+
+        // try to push register if we have enough buffer.
+        // No other checks for the same reason as the above step.
+        #[cfg(feature = "registers")]
```

Comment:
the following block of code can be inside the same above block, under one feature guard ?